src/sage/geometry/hyperplane_arrangement/arrangement.py

```diff
@@ -3596,6 +3663,395 @@ def derivation_module_basis(self, algorithm="singular"):
+        if proj and not self.is_central():
+            raise TypeError('The arrangement is not central')
+        n0 = self.dimension()
+        if not proj:
+            H = self.cone()
+            H1 = H.hyperplane_section()
+            mat = Matrix(h.coefficients()[1:] for h in H1)
+            m = mat.nrows()
+            for j in range(mat.ncols()):
+                if mat[m - 1, j] != 0:
+                    mat.swap_columns(0, j)
+                    break
+            for j in range(1, mat.ncols()):
+                mat.add_multiple_of_column(j, 0, -mat[m - 1, j] / mat[m - 1, 0])
+            vrs = H1.parent().variable_names()[1:]
+            A1 = OrderedHyperplaneArrangements(self.base_ring(), names=vrs)
+            mat_rows = mat.rows()[:-1]
+            H1b = A1(mat_rows)
+            return H1b
+        P = self.intersection_poset(element_label="subspace")
+        n1 = self.center().dimension()
+        U = [p.linear_part().basis()[0] for p in P if p.dimension() == n1 + 1]
+        U0 = sum(U)
+        for v in ZZ**n0:
+            v1 = v + U0
+            if 0 not in [w * v1 for w in U]:
+                break
+        h0 = self.parent()((0,) + tuple(v1))
+        H1 = self.add_hyperplane(h0)
+        return H1.restriction(h0)
+
+    def affine_fundamental_group(self):
+        r"""
+        It computes the fundamental group of the complement of an affine
+        hyperplane arrangement in `\mathbb{C}^n` whose equations have
+        coefficients in a subfield of `\overline{\mathbb{Q}}`
+
+        OUTPUT:
+
+        A finitely presented group.
+
+        EXAMPLES::
+
+            sage: # needs sirocco
+            sage: A.<x, y> = OrderedHyperplaneArrangements(QQ)
+            sage: L = [y + x, y + x - 1]
+            sage: H = A(L)
+            sage: H.affine_fundamental_group()
+            Finitely presented group < x0, x1 |  >
+            sage: L = [x, y, x + 1, y + 1, x - y]
+            sage: A(L).affine_fundamental_group()
+            Finitely presented group
+            < x0, x1, x2, x3, x4 | x4*x0*x4^-1*x0^-1,
+                                   x0*x2*x3*x2^-1*x0^-1*x3^-1,
+                                   x1*x2*x4*x2^-1*x1^-1*x4^-1,
+                                   x2*x3*x0*x2^-1*x0^-1*x3^-1,
+                                   x2*x4*x1*x2^-1*x1^-1*x4^-1,
+                                   x4*x1*x4^-1*x3^-1*x2^-1*x1^-1*x2*x3 >
+            sage: H = A(x, y, x + y)
+            sage: H.affine_fundamental_group()
+            Finitely presented group
+            < x0, x1, x2 | x0*x1*x2*x1^-1*x0^-1*x2^-1, x1*x2*x0*x1^-1*x0^-1*x2^-1 >
+
+        .. WARNING::
+
+            This functionality requires the sirocco package to be installed.
+        """
+        K = self.base_ring()
+        if not K.is_subring(QQbar):
+            raise TypeError('the base field is not in QQbar')
+        if self._affine_fundamental_group_:
+            return self._affine_fundamental_group_
```

Comment:
I was advised to use it and in the meantime I found it useful, for example to port computations to another session. Anyway i take out the ending underscore.

Reply:
I thought pickling worked just fine with `@cached_method` as it stores stuff in the `__dict__`, which should be put into the pickle:
```
sage: class Foo:
....:     @cached_method
....:     def bar(self, x):
....:         return x
....:
sage: F = Foo
sage: F = Foo()
sage: F.__dict__
{}
sage: F.bar(2)
2
sage: F.__dict__
{'_cache__bar': {((2,), ()): 2},
 'bar': Cached version of <function Foo.bar at 0x7f1a12741750>}
```
and thus work across sessions. Using `@cached_method` is also faster since it is implemented in Cython and it results it cleaner code. I strongly recommend you to use it where appropriate.

Reply:
Let me try next week. Does it means that if I save an object in the class I can recover it with `__dict__`?

Reply:
I made a test but it is not preserved by saving the object and loading it in another session

Reply:
You need to add one other option `do_pickle=True` to the decorator (read the `cached_method` doc). However, I would be hesitant with wanting computed information to be put into pickles to translate across sessions. I find it a better practice to manually store that information and pass it accordingly. Less data in the pickles and more likely to catch bugs.

Reply:
I understand what you mean, and this is usually what I do. The advantage of attributes or pickle is that one saves just one object. On the other side, to save manually the group I had strange issues, and the better way was to save the number of generators and the Tietze words of the relations.
Actually I was trying it adding `do_pickle=True` for the methods `fundamental_group` and `braid_monodromy` for affine plane curves. Once the change has been done I run:
```
sage: A.<x,y> = AffineSpace(QQ, 2)
sage: C = Curve(y^2 -x^3)
sage: bm = C.braid_monodromy()
sage: g = C.fundamental_group()
sage: save(C, 'aa')
```
Everything works fine, but starting another session:
```
sage: C = load('aa.sobj')
File /usr/local/sage-310/src/sage/misc/persist.pyx:189, in sage.misc.persist.load()
    187 
    188     ## Load file by absolute filename
--> 189     with open(filename, 'rb') as fobj:
    190         X = loads(fobj.read(), compress=compress, **kwargs)
    191     try:

File /usr/local/sage-310/src/sage/misc/persist.pyx:190, in sage.misc.persist.load()
    188 ## Load file by absolute filename
    189 with open(filename, 'rb') as fobj:
--> 190     X = loads(fobj.read(), compress=compress, **kwargs)
    191 try:
    192     X._default_filename = os.path.abspath(filename)

File /usr/local/sage-310/src/sage/misc/persist.pyx:995, in sage.misc.persist.loads()
    993 unpickler = SageUnpickler(io.BytesIO(s), **kwargs)
    994 global already_unpickled
--> 995 ans = unpickler.load()
    996 already_unpickled = { }
    997 return ans

File /usr/local/sage-310/src/sage/libs/gap/element.pyx:390, in sage.libs.gap.element._from_sage()
    388 
    389 
--> 390 cpdef _from_sage(elem) noexcept:
    391     """
    392     Currently just used for unpickling; equivalent to calling ``libgap(elem)``

File /usr/local/sage-310/src/sage/libs/gap/element.pyx:396, in sage.libs.gap.element._from_sage()
    394 """
    395 if isinstance(elem, str):
--> 396     return libgap.eval(elem)
    397 
    398 return libgap(elem)

File /usr/local/sage-310/src/sage/libs/gap/libgap.pyx:406, in sage.libs.gap.libgap.Gap.eval()
    404 
    405         initialize()
--> 406         elem = make_any_gap_element(self, gap_eval(gap_command))
    407 
    408         # If the element is NULL just return None instead

File /usr/local/sage-310/src/sage/libs/gap/util.pyx:364, in sage.libs.gap.util.gap_eval()
    362 try:
    363     GAP_Enter()
--> 364     result = GAP_EvalString(cmd)
    365     # We can assume that the result object is a GAP PList (plain list)
    366     # and we should use functions for PLists directly for now; see

GAPError: Error, Variable: 'x0' must have a value
```